applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp

```diff
@@ -0,0 +1,629 @@
+        mSections[i]->FinalizeSolutionStep(props, geom, row(shapeFunctionsValues, i), rCurrentProcessInfo);
+}
+
+void BaseShellElement::CalculateLocalSystem(MatrixType& rLeftHandSideMatrix,
+	VectorType& rRightHandSideVector,
+	ProcessInfo& rCurrentProcessInfo)
+{
+    // Calculation flags
+    const bool calculate_stiffness_matrix_flag = true;
+    const bool calculate_residual_vector_flag = true;
+
+	CalculateAll(rLeftHandSideMatrix, rRightHandSideVector, rCurrentProcessInfo,
+                 calculate_stiffness_matrix_flag, calculate_residual_vector_flag);
+}
+
+void BaseShellElement::CalculateRightHandSide(VectorType& rRightHandSideVector,
+	ProcessInfo& rCurrentProcessInfo)
+{
+    // Calculation flags
+    const bool calculate_stiffness_matrix_flag = true; // TODO check is this can be false => see solids
+    const bool calculate_residual_vector_flag = true;
+
+	Matrix dummy;
+	CalculateAll(dummy, rRightHandSideVector, rCurrentProcessInfo,
+                 calculate_stiffness_matrix_flag, calculate_residual_vector_flag);
+}
+
+void BaseShellElement::CalculateMassMatrix(MatrixType& rMassMatrix, ProcessInfo& rCurrentProcessInfo)
+{
+    // TODO unify implementation and move it to BaseClass
+}
+
+void BaseShellElement::CalculateDampingMatrix(
+    MatrixType& rDampingMatrix,
+    ProcessInfo& rCurrentProcessInfo
+    )
+{
+    KRATOS_TRY;
+
+    if ( rDampingMatrix.size1() != mNumDofs )
+        rDampingMatrix.resize( mNumDofs, mNumDofs, false );
+
+    noalias( rDampingMatrix ) = ZeroMatrix( mNumDofs, mNumDofs );
+
+    // 1.-Calculate StiffnessMatrix:
+
+    MatrixType StiffnessMatrix  = Matrix();
+    VectorType ResidualVector  = Vector();
+
+    CalculateAll(StiffnessMatrix, ResidualVector, rCurrentProcessInfo, true, false);
+
+    // 2.-Calculate MassMatrix:
+
+    MatrixType MassMatrix  = Matrix();
+
+    CalculateMassMatrix(MassMatrix, rCurrentProcessInfo);
+
+    // 3.-Get Damping Coeffitients (RAYLEIGH_ALPHA, RAYLEIGH_BETA)
+    double alpha = 0.0;
+    if( GetProperties().Has(RAYLEIGH_ALPHA) )
+        alpha = GetProperties()[RAYLEIGH_ALPHA];
+    else if( rCurrentProcessInfo.Has(RAYLEIGH_ALPHA) )
+        alpha = rCurrentProcessInfo[RAYLEIGH_ALPHA];
+
+    double beta  = 0.0;
+    if( GetProperties().Has(RAYLEIGH_BETA) )
+        beta = GetProperties()[RAYLEIGH_BETA];
+    else if( rCurrentProcessInfo.Has(RAYLEIGH_BETA) )
+        beta = rCurrentProcessInfo[RAYLEIGH_BETA];
+
+    // 4.-Compose the Damping Matrix:
+
+    // Rayleigh Damping Matrix: alpha*M + beta*K
+    noalias( rDampingMatrix ) += alpha * MassMatrix;
+    noalias( rDampingMatrix ) += beta  * StiffnessMatrix;
+
+    KRATOS_CATCH( "" )
+}
+
+int BaseShellElement::Check(const ProcessInfo& rCurrentProcessInfo)
+{
+    KRATOS_TRY;
+
+    Element::Check(rCurrentProcessInfo);
+
+    CheckVariables();
+    CheckDofs();
+    CheckProperties(rCurrentProcessInfo);
+
+    KRATOS_ERROR_IF(GetGeometry().Area() < std::numeric_limits<double>::epsilon()*1000)
```

Comment:
It's a good practise to consider epsilon*1000?

Reply:
IDK, I saw it somewhere recently
But in general I am still not decided how to use it
I also thought of using 1e-15

Reply:
i personally prefer the 1000*eps.